kwok/ec2/ec2.go

```diff
@@ -0,0 +1,890 @@
+		if err != nil {
+			return false
+		}
+		return iNum < jNum
+	})
+	// Clean-up any ConfigMaps that don't need to be there because of the count
+	// We store 500 instances per ConfigMap
+	numConfigMaps := int(math.Ceil(float64(len(instances)) / float64(500)))
+	if numConfigMaps < len(configMaps.Items) {
+		errs := make([]error, numConfigMaps)
+		workqueue.ParallelizeUntil(ctx, 10, len(configMaps.Items)-numConfigMaps, func(i int) {
+			if err := c.kubeClient.Delete(ctx, &configMaps.Items[len(configMaps.Items)-i-1]); client.IgnoreNotFound(err) != nil {
+				errs[i] = fmt.Errorf("deleting configmap %q, %w", configMaps.Items[len(configMaps.Items)-i-1].Name, err)
+			}
+		})
+		if err := multierr.Combine(errs...); err != nil {
+			return err
+		}
+	}
+
+	errs := make([]error, numConfigMaps)
+	workqueue.ParallelizeUntil(ctx, 10, numConfigMaps, func(i int) {
+		cm := &corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      fmt.Sprintf("kwok-aws-instances-%d", i),
+				Namespace: "kube-system",
+			},
+		}
+		if err := c.kubeClient.Get(ctx, client.ObjectKeyFromObject(cm), cm); err != nil {
+			if k8serrors.IsNotFound(err) {
+				if err = c.kubeClient.Create(ctx, cm); err != nil {
+					errs[i] = fmt.Errorf("creating configmap %q, %w", cm.Name, err)
+					return
+				}
+			} else {
+				errs[i] = fmt.Errorf("getting configmap %q, %w", cm.Name, err)
+				return
+			}
+		}
+		stored := cm.DeepCopy()
+		cm.Data = map[string]string{"instances": string(removeNullFields(lo.Must(json.Marshal(lo.Slice(instances, i*500, (i+1)*500)))))}
+		if !equality.Semantic.DeepEqual(cm, stored) {
+			if err := c.kubeClient.Patch(ctx, cm, client.MergeFrom(stored)); err != nil {
+				errs[i] = fmt.Errorf("patching configmap %q, %w", cm.Name, err)
+				return
+			}
+		}
+	})
+	return multierr.Combine(errs...)
+}
+
+// StartBackupThread initiates the thread that is responsible for storing instances in ConfigMaps on the cluster
+func (c *Client) StartBackupThread(ctx context.Context) {
+	for {
+		if err := c.backupInstances(ctx); err != nil {
+			log.FromContext(ctx).Error(err, "unable to backup instances")
+		}
+		select {
+		case <-time.After(time.Second * 5):
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// StartKillNodeThread initiates the thread that is responsible for killing nodes on the cluster that no longer have an instance representation (similar to CCM)
+func (c *Client) StartKillNodeThread(ctx context.Context) {
+	for {
+		nodes := &corev1.NodeList{}
+		if err := c.kubeClient.List(ctx, nodes, client.MatchingLabels{v1alpha1.KwokLabelKey: v1alpha1.KwokLabelValue}); err != nil {
+			log.FromContext(ctx).Error(err, "unable to list nodes")
+			continue
+		}
+		for _, node := range nodes.Items {
+			id, err := utils.ParseInstanceID(node.Spec.ProviderID)
+			if err != nil {
+				log.FromContext(ctx).Error(err, "unable to parse instance id for node %q", node.Name)
+				continue
+			}
+			if _, ok := c.instances.Load(id); !ok {
```

Comment:
Want to use multiple goroutines? Should be fine to go super fast with KWOK.

Reply:
Seems like it would be a pretty minimal gain to add this -- we can always add this later if we think it's too slow